lib/fizzy/parser.cpp

```diff
@@ -275,14 +276,41 @@ Module parse(bytes_view input)
         }
 
         if (it != expected_end_pos)
+        {
             throw parser_error{"incorrect section " + std::to_string(static_cast<int>(id)) +
                                " size, difference: " + std::to_string(it - expected_end_pos)};
+        }
+    }
+
+    if (module.tablesec.size() > 1)
+        throw parser_error{"too many table sections (at most one is allowed)"};
+
+    if (module.memorysec.size() > 1)
+        throw parser_error{"too many memory sections (at most one is allowed)"};
+
+    const auto imported_memories = std::count_if(module.importsec.begin(), module.importsec.end(),
+        [](const auto& import) noexcept { return import.kind == ExternalKind::Memory; });
 
-        if (module.tablesec.size() > 1)
-            throw parser_error{"too many table sections (at most one is allowed)"};
+    if (imported_memories > 1)
+        throw parser_error{"too many imported memories (at most one is allowed)"};
 
-        if (module.memorysec.size() > 1)
-            throw parser_error{"too many memory sections (at most one is allowed)"};
+    if (!module.memorysec.empty() && imported_memories > 0)
```

Comment:
Why not `module.memorysec.size() + imported_memories <= 1` since we ensure `memorysec.size() <= 1` ?

Reply:
Just becasue `empty()` presumably could be more efficient in some implementations than `size()`. But here probably it's not that important, perhaps your variant looks nicer.

Reply:
It would require a cast, so won't look particularly nice either
```suggestion
    if (module.memorysec.size() + static_cast<size_t>(imported_memories) > 1)
```